src/algebra/group/regular.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+
+/-!
+# Regular elements
+
+We introduce left-regular, right-regular and regular elements.  The final goal is to develop
+part of the API to prove, eventually, results about non-zero-divisors.
+
+-/
+variables {R : Type*}
+
+/-- A regular element is an element `c` such that multiplication by `c` both on the left and
+on the right is injective. -/
+structure is_regular [has_mul R] (c : R) : Prop :=
+( mul_left_inj : ∀ a b : R, c * a = c * b → a = b)
+( mul_right_inj : ∀ a b : R, a * c = b * c → a = b)
+
+namespace is_regular
+
+/-- A left-regular element is an element `c` such that multiplication on the left by `c`
+is injective on the left. -/
+structure left [has_mul R] (c : R) : Prop :=
+( mul_left_inj : ∀ a b : R, c * a = c * b → a = b)
+
+/-- A right-regular element is an element `c` such that multiplication on the right by `c`
+is injective on the left. -/
+structure right [has_mul R] (c : R) : Prop :=
+( mul_right_inj : ∀ a b : R, a * c = b * c → a = b)
+
```

Comment:
Note that these definitions of `left` and `right` do not yet give you a projection function `is_regular c -> is_regular.left c`.
It might make sense to return to `is_left_regular` and have `is_regular.left` be that projection function.

Reply:
I removed the spaces and reverted to the old names.

However, I do not understand what this projection thing is...

Reply:
Should I define a lemma called `left` in the `is_regular` namespace that assumes `is_regular` and proves `is_left_regular`?

Reply:
Yes, that sounds good